### PR TITLE
test(optimizer):add test for distributed plan

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Review by chenzl25 — APPROVED** [review]
LGTM

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml [hunk 1]
```diff
@@ -1,63 +1,429 @@
...
+      StreamExchange { dist: HashShard(ak1.a._row_id, b._row_id) }
+        StreamHashJoin { type: Inner, predicate: ak1.k1 = b.k1, output: [ak1.v, b.v, ak1.a._row_id, b._row_id] }
+          StreamExchange { dist: HashShard(ak1.k1) }
+            StreamTableScan { table: ak1, columns: [ak1.k1, ak1.v, ak1.a._row_id], pk: [ak1.a._row_id], distribution: UpstreamHashShard(ak1.k1) }
+          StreamExchange { dist: HashShard(b.k1) }
+            StreamTableScan { table: b, columns: [b.k1, b.v, b._row_id], pk: [b._row_id], distribution: UpstreamHashShard(b._row_id) }
+- id: A_join_Bk1_onk1
+  before:
+  - create_tables
+  sql: select A.v, B.v as Bv from Ak1 as A join Bk1 as B using(k1)
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/distribution_derive.yaml:48]:
```suggestion
  sql: select A.v, B.v as Bv from A join Bk1 as B using(k1)
```

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml [hunk 2]
```diff
@@ -1,63 +1,429 @@
...
+            BatchScan { table: ak1k2, columns: [ak1k2.k2, ak1k2.v], distribution: SomeShard }
+  stream_plan: |
+    StreamMaterialize { columns: [max_v, ak1k2.k2(hidden)], pk_columns: [ak1k2.k2] }
+      StreamProject { exprs: [max(max(ak1k2.v)), ak1k2.k2] }
+        StreamHashAgg { group_key: [ak1k2.k2], aggs: [sum(count), max(max(ak1k2.v))] }
+          StreamExchange { dist: HashShard(ak1k2.k2) }
+            StreamHashAgg { group_key: [ak1k2.k2, Vnode(ak1k2.k1, ak1k2.k2)], aggs: [count, max(ak1k2.v)] }
+              StreamProject { exprs: [ak1k2.k2, ak1k2.v, ak1k2.k1, ak1k2.a._row_id, Vnode(ak1k2.k1, ak1k2.k2)] }
+                StreamTableScan { table: ak1k2, columns: [ak1k2.k2, ak1k2.v, ak1k2.k1, ak1k2.a._row_id], pk: [ak1k2.a._row_id], distribution: UpstreamHashShard(ak1k2.k1, ak1k2.k2) }
+- id: aggk1k2_from_Ak1k2
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/distribution_derive.yaml:165]: Add another `aggk1k2_from_Ak1`?

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml [hunk 3]
```diff
@@ -1,63 +1,429 @@
...
+  sql: |
+    select
+      max(num) as max_num
+    from (
+      select
+        count(*) as num, k2
+      from A
+      group by k1, k2
+    )
+    group by k2;
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/distribution_derive.yaml:298]:
```suggestion
        count(*) as num, k1, k2
      from A
      group by k1, k2
    )
    group by k1, k2;
```
?

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml [hunk 4]
```diff
@@ -1,63 +1,429 @@
...
+  before:
+  - create_tables
+  sql: |
+    with b as (
+      select
+        count(*) as num, k1
+      from A
+      group by k1
+    )
+    select A.v, B.num as Bv from Ak1 as A join B using(k1)
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/distribution_derive.yaml:331]:
```suggestion
    with aggk1 as (
      select
        count(*) as num, k1
      from A
      group by k1
    )
    select A.v, B.num as Bv from Ak1 as A join aggk1 as B using(k1)
```

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml [hunk 5]
```diff
@@ -1,63 +1,429 @@
...
+  before:
+  - create_tables
+  sql: |
+    with b as (
+      select
+        count(*) as num, k1
+      from A
+      group by k1
+    )
+    select A.v, B.num as Bv from B join Ak1 as A using(k1)
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/distribution_derive.yaml:363]:
```suggestion
    with aggk1 as (
      select
        count(*) as num, k1
      from A
      group by k1
    )
    select A.v, B.num as Bv from aggk1 as B join Ak1 as A using(k1)
```

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml [hunk 6]
```diff
@@ -1,63 +1,429 @@
...
+        count(*) as num, k1
+      from A
+      group by k1
+    ), b as (
+      select
+        count(*) as num, k1
+      from B
+      group by k1
+    )
+    select A.num, B.num as Bv from a join b using(k1)
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/distribution_derive.yaml:399]:
```suggestion
    select a.num, b.num as Bv from a join b using(k1)
```

Comment (xxchan) [reply src/frontend/test_runner/tests/testdata/distribution_derive.yaml:399]: BTW, can `aggk1 as A join aggk1 as B`

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml (hunk 3 above)
Comment (st1page) [reply src/frontend/test_runner/tests/testdata/distribution_derive.yaml:298]: just use another unify table defination

File: src/frontend/test_runner/tests/testdata/distribution_derive.yaml (hunk 6 above)
Comment (st1page) [reply src/frontend/test_runner/tests/testdata/distribution_derive.yaml:399]: there is some bugs, I will fix it in another PR

Comment (xxchan) [reply src/frontend/test_runner/tests/testdata/distribution_derive.yaml:399]: Why not fix them here 😢

**Pull request merged**